### ISSUE TITLE
Documentation around TimeoutStopSec= behaviour is not clear

### ISSUE THREAD

**Issue by amiryal** [systemd/systemd#5773]
### Submission type

Bug report (in documentation)

### systemd version the issue has been seen with

233

### In case of bug report: Unexpected behaviour you saw

https://www.freedesktop.org/software/systemd/man/systemd.service.html#TimeoutStopSec= says:
> If a service is asked to stop, but does not terminate **in the specified time**, it will be terminated forcibly via SIGTERM, and after another timeout of equal duration with SIGKILL

https://www.freedesktop.org/software/systemd/man/systemd.service.html#ExecStop= says:
> Since the remaining processes of the services are killed according to KillMode= and KillSignal= as described above **immediately** after the command exited, this may not result in a clean stop

The two messages (highlighted in bold by me) are conflicting. When requested to stop a service, does systemd send SIGTERM immediately after ExecStop, or does it wait for TimeoutStopSec to elapse first?

Once that question has been answered, the manual page should be amended to be consistent with the answer.

Poking around [the source code](https://github.com/systemd/systemd/blob/74278ce56732056bb3cc75062113e393db1d6dd2/src/core/service.c#L1650) a little, I get the impression that the latter is correct – i.e. SIGTERM following a ExecStop command is sent only after a timeout, not immediately – but please double-check.

Thanks,
Amir

**Comment by arvidjaar** [comment]
I do not think messages are conflicting. The former describes what happens when command did not finish before timeout ("does not terminate in specified time") and the latter describes what happens when command did finish before timeout ("after command exited"). Probably this needs clarification what exactly "service did not terminate in time" means though.

**Comment by boucman** [comment]
that, and where exactly is TimeoutStopSec. IIUC, the sequence of event is

1. ExecStop
1. SigTerm (or whatever you replace it with
1. SigKill

I'm not sure where exactly is the TimeoutStopSec applied. between 1 and 2 or 1 and 2 and a second time with 2 and 3

anyway, yes, that needs clarification

**Comment by arvidjaar** [comment]
> I'm not sure where exactly is the TimeoutStopSec applied.

AFAICT it is applied to each command/kill invocation, including those initiated by timeout. So if you have `ExecStop` and `ExecStopPost`  then total time may be `6 * TimeoutStopSec` - main stop command, probably `SIGTERM` and `SIGKILL` if main stop command timeouts (each with `TimeoutStop`), post command, final `SIGTERM` and `SIGKILL`.

This is complicated by the fact that timeout during one `ExecStopPost` command will skip all further `ExecStopPost` commands and immediately proceed to final `SIGTERM`.

**Comment by keszybz** [comment]
@boucman the two quotes in your original description talk about different classes of processes. The first is about the controlling process, the second is about any processes remaining after the controlling processes has exited (on its own, or by signal). The timeouts apply to the first class, and not to the second class (any processes remaining after the main process has exited is an error in the service). The second quote states that pretty clearly, I don't see anything wrong or unclear in that description.

**Comment by poettering** [comment]
While i don't think the two bits are contradicting, i figure we could reword this to clear up any potential confusion. After all, it appears that it did result in confusion, given that @amiryal posted a bug about it...

**Comment by amiryal** [comment]
Maybe change (in the paragraph describing `ExecStop=`) **from** “After the commands configured in this option are run, all processes remaining for a service are terminated [...]” **to** “After the commands configured in this option are run, it is implied that the service is stopped, and any processes remaining for it are terminated [...]”.
